### PR TITLE
LF: Add rank to ValueVariant and ValueEnum

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/protobuf/com/daml/lf/value.proto [hunk 1]
```diff
@@ -101,12 +101,14 @@ message Identifier {
...
     Identifier variant_id = 1; // optional
     string constructor = 2;
+    uint32 rank = 4; // available in >= 12
     Value value = 3;
 }
 
 message Enum {
     Identifier enum_id = 1; // optional
     string value = 2;
+    uint32 rank = 4; // available in >= 12
```
Comment (stefanobaghino-da) [inline daml-lf/transaction/src/main/protobuf/com/daml/lf/value.proto:111]: Why not `3` here?

**Review by stefanobaghino-da — COMMENTED** [review]
Does this have an impact on the serialization benchmarks?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 2]
```diff
@@ -90,8 +90,8 @@ sealed trait SValue {
         SRecord(tycon, fields, mapArrayList(values, v => v.mapContractId(f)))
       case SStruct(fields, values) =>
         SStruct(fields, mapArrayList(values, v => v.mapContractId(f)))
-      case SVariant(tycon, variant, rank, value) =>
-        SVariant(tycon, variant, rank, value.mapContractId(f))
+      case SVariant(tycon, rank, variant, value) =>
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:93]: I’m confused, here rank is before variant and a few lines above it’s the other way around? One of those must be off

File: daml-lf/transaction-test-lib/src/main/scala/lf/value/test/TypedValueGenerators.scala [hunk 3]
```diff
@@ -416,7 +417,10 @@ object TypedValueGenerators {
         : Arbitrary[HRec[Cid]]
     private[TypedValueGenerators] implicit def recshrink[Cid: Shrink]: Shrink[HRec[Cid]]
 
-    private[TypedValueGenerators] def injVar[Cid: IntroCtx](v: HVar[Cid]): (Ref.Name, Value[Cid])
+    private[TypedValueGenerators] def injVar[Cid: IntroCtx](
+        v: HVar[Cid],
+        rank: Int = 0,
```
Comment (cocreature) [inline daml-lf/transaction-test-lib/src/main/scala/lf/value/test/TypedValueGenerators.scala:422]: is the default here useful? this seems like it’s just asking for incorrect ranks.

File: daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueNormalizer.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.daml.lf.value.test
+
+import com.daml.lf.transaction.TransactionVersion
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value._
+
+object ValueNormalizer {
+
+  // equivalent to serialization + unserialization.
```
Comment (cocreature) [inline daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueNormalizer.scala:12]: this only normalizes ranks, should it also normalize constructors?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala [hunk 5]
```diff
@@ -0,0 +1,344 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:1]: this seems like it is somewhat overlapping with #7824 so unless that is incorrect, I’ll review this once we’ve merged the other one.

**Review by cocreature — COMMENTED** [review]
lgtm, I’ll take another detailed look once this is rebased on the type constructor changes.

File: daml-lf/transaction-test-lib/src/main/scala/lf/value/test/TypedValueGenerators.scala (hunk 3 above)
Comment (S11001001) [reply daml-lf/transaction-test-lib/src/main/scala/lf/value/test/TypedValueGenerators.scala:422]: Indeed it's asking for things to go wrong.

**Pull request closed without merging**